pybamm/models/submodels/pe_degradation/base_phase_transition.py

```diff
@@ -0,0 +1,265 @@
+            "oxygen concentration [mol.m-3]": c_o_xav,
+            f"R-averaged {domain} {phase_name}shell "
+            "oxygen concentration [mol.m-3]": c_o_rav,
+            f"Average {domain} {phase_name}shell "
+            "oxygen concentration [mol.m-3]": c_o_av,
+            f"{Domain} {phase_name}shell inner boundary "
+            "oxygen concentration [mol.m-3]": c_o_inner,
+            f"X-averaged {domain} {phase_name}shell inner boundary "
+            "oxygen concentration [mol.m-3]": c_o_inner_av,
+            f"{Domain} {phase_name}shell surface "
+            "oxygen concentration [mol.m-3]": c_o_surf,
+            f"X-averaged {domain} {phase_name}shell surface"
+            "oxygen concentration [mol.m-3]": c_o_surf_av,
+            # Core-shell phase boundary
+            f"{Domain} {phase_name}particle moving phase boundary location" "": s_nd,
+            f"X-averaged {domain} {phase_name}particle "
+            "moving phase boundary location": s_nd_av,
+            # Boundary cell values
+            f"{Domain} {phase_name}core surface cell "
+            "lithium concentration [mol.m-3]": c_c_N,
+            f"X-averaged {domain} {phase_name}core surface cell "
+            "lithium concentration [mol.m-3]": c_c_N_av,
+            f"{Domain} {phase_name}shell inner cell "
+            "oxygen concentration [mol.m-3]": c_o_1,
+            f"X-averaged {domain} {phase_name}shell inner cell "
+            "oxygen concentration [mol.m-3]": c_o_1_av,
+            # Boundary half cell length (central node to edge)
+            f"{Domain} {phase_name}core surface cell length [m]": dx_core,
+            f"X-averaged {domain} {phase_name}core surface cell length [m]"
+            "": dx_core_av,
+            f"{Domain} {phase_name}shell inner cell length [m]": dx_shell,
+            f"X-averaged {domain} {phase_name}shell inner cell length [m]"
+            "": dx_shell_av,
+            # Loss of active material (LAM) due to progressing of s
+            # the shell is considered as LAM
+            "Loss of active material due to PE phase transition": lam_pe,
+            "X-averaged loss of active material due to PE phase transition"
+            "": lam_pe_av,
+            # ---------------------------------------------------------------------
+            # We need to supply a variable to
+            # submodel 'positive interface' (set_interfacial_submodel)
+            # for voltage calculation
+            f"{Domain} {phase_name}particle surface concentration [mol.m-3]"
+            "": c_c_surf,
+            # For submodel 'positive primary open-circuit potential'
+            # f"{Domain} {phase_name}core surface lithium stoichiometry"
+            f"{Domain} {phase_name}particle surface stoichiometry" "": c_c_surf
+            / c_scale,
+            # ---------------------------------------------------------------------
+        }
+
+        return variables
+
+    def _r_average_core(self, symbol):
```

Comment:
again, I think this kind of thing should be handled by the spatial methods code, rather than explicitly as part of the model

Reply:
Hi Rob, thanks for reviewing the codes. Regarding your main concern, the PE degradation submodel consists of two diffusion equations residing in a shrinking core and expanding shell, and a set of ODEs describing the moving of the phase boundary. The first step is to do variable change so that the computational domains (core and shell) are fixed, this will naturally change the governing diffusion equations by adding extra terms associated with the moving boundary, which is an unknown variable. The extra terms are caused by both the original temporal differential term and spatial differential terms, so it does not make too much sense to incorporate into the discretization method. Another important reason for me doing the current way is to minimize the modification to current pybamm, so I leave the first step to manual calculation. To be honest, I have no idea whether this can be incorporated into the discretized method, or if yes, how much cost it will take. Regrading the [0, 1] issues, that is how I did in the initial version, now to be consistent with current pybamm release, I scaled both up with a constant R_typ. I share the paper and the formula for implementation: https://doi.org/10.1016/j.jpowsour.2022.232461 and https://hkustconnect-my.sharepoint.com/:b:/g/personal/mzhuo_connect_ust_hk/EYmFXnyOpP9KlUsrHg_0RtIBqQoIkN6qIpVCKu262uSgLA?e=EH0SJX